### ISSUE TITLE
Time slice and annotation references

### ISSUE THREAD

**Issue by achilleas-k** [NeuralEnsemble/python-neo#536]
This issue came up while I was investigating some of the details behind issue #535.

A quick reminder of a design decision made in issue #311: The NixIO uses object annotations to store names for identifying objects. This is especially useful for objects that are referenced in multiple locations and for references that don't have a `name`.

In the code posted in issue #535, an AnalogSignal gets sliced and then the sliced signals get stored. I noticed this was creating strange behaviour. The sliced signals reference the same `annotations` as the original, which means once the NixIO writes an annotation to give the signal a name, all the signals get the same name.

```python
import neo
import quantities as pq


asig = neo.AnalogSignal([1, 2, 10], units="mV", sampling_rate=pq.Hz)
asig.annotate(somesortofkey="annotation value")
print(asig.annotations)
print(id(asig.annotations))

slicea = asig.time_slice(t_start=1*pq.s, t_stop=2*pq.s)
print(slicea.annotations)
print(id(slicea.annotations))

sliceb = asig.time_slice(t_start=2*pq.s, t_stop=3*pq.s)
print(sliceb.annotations)
print(id(sliceb.annotations))
```

Output:
```
{'somesortofkey': 'annotation value'}
140606369415960
{'somesortofkey': 'annotation value'}
140606369415960
{'somesortofkey': 'annotation value'}
140606369415960
```

Is this intended behaviour? If it's not, the `time_slice` method should be doing a deeper copy.
If it is, I need to redesign the object naming in the NixIO. As it is now, the two slices that share the same annotations wont be stored, since they all look like the same object during the write operation.

**Comment by grg2rsr** [comment]
Do you know if there is a current workaround until the WIP PR is done and merged?

**Comment by achilleas-k** [comment]
The most straightforward solution is to reinitialise or copy the annotations after slicing.

In your example from the #535 that would be:
```python
for t in Blk.segments[0].events[0].times:
    segment = neo.core.Segment()
    Asig = AnalogSignal.time_slice(*(t+t_cut))
    # reinitialise annotations
    Asig.annotations = dict()
    # annotate
    Asig.annotate(kind='original')
    # append to Segment
    segment.analogsignals.append(Asig)
    Segments.append(segment)
```

If you want to copy the annotations from the original signal, you can do the following:
```python
for t in Blk.segments[0].events[0].times:
    segment = neo.core.Segment()
    Asig = AnalogSignal.time_slice(*(t+t_cut))
    # copy annotations from original signal
    Asig.annotations = AnalogSignal.annotations.copy()
    del Asig.annotations["nix_name"]
    # annotate
    Asig.annotate(kind='original')
    # append to Segment
    segment.analogsignals.append(Asig)
    Segments.append(segment)
```

I'm not sure if the second one is a good solution. It might create issues because it's a shallow copy, so modifying an annotation in place would be modifying every reference to it in all the other objects. Julia's PR (#538) does a deep copy which solves this issue.

This also got me thinking that the NixIO would need some modifications still, even after PR #538 lands, since the `nix_name` will get copied to all signal slices. I have to rethink certain details of the identification of objects.

**Comment by apdavison** [comment]
@achilleas-k what is the current status of this issue?

**Comment by achilleas-k** [comment]
Looks like it was fixed in #668.  I guess I missed that.  Closing now.
